### PR TITLE
Add guiderate for injection groups

### PR DESCRIPTION
**Pull request by totto82**
This adapt the guiderate classes to handle injection groups. It needs to be merge together with downstream changes in opm-simulators.

### CODE REVIEW

**Comment by totto82** [comment]
jenkins build this opm-simulators=3112 please

File: opm/parser/eclipse/EclipseState/Schedule/Group/GuideRate.hpp
Comment (joakim-hove) [inline opm/parser/eclipse/EclipseState/Schedule/Group/GuideRate.hpp:98]: I would prefer if the short `get( )` name was retained - the well/injection group/produvtion group can be differntiated by argument overloading.

Comment (joakim-hove) [inline opm/parser/eclipse/EclipseState/Schedule/Group/GuideRate.hpp:98]: The arguments  should be reordered -  the groupname is the most important argument and should come first.

File: opm/parser/eclipse/EclipseState/Schedule/Group/GuideRate.hpp
Comment (joakim-hove) [inline opm/parser/eclipse/EclipseState/Schedule/Group/GuideRate.hpp:102]: Again - name `compute()` plus overload should be enough.

File: opm/parser/eclipse/EclipseState/Schedule/Group/GuideRate.hpp
Comment (joakim-hove) [inline opm/parser/eclipse/EclipseState/Schedule/Group/GuideRate.hpp:126]: I would prefer: `std::unordered_map<std::string, std::unordered_map<Phase, double>>`

File: src/opm/parser/eclipse/EclipseState/Schedule/Group/GuideRate.cpp
Comment (joakim-hove) [inline src/opm/parser/eclipse/EclipseState/Schedule/Group/GuideRate.cpp:111]:
```C++
auto messag = fmt::format("Did not find .... group {}:{}", name, std::to_string(...));
```

File: src/opm/parser/eclipse/EclipseState/Schedule/Group/GuideRateConfig.cpp
Comment (joakim-hove) [inline src/opm/parser/eclipse/EclipseState/Schedule/Group/GuideRateConfig.cpp:90]: I guess `properties`could be `const`?

File: opm/parser/eclipse/EclipseState/Schedule/Group/GuideRate.hpp
Comment (totto82) [reply opm/parser/eclipse/EclipseState/Schedule/Group/GuideRate.hpp:126]: I know. I tried that as well. But I got lost somewhere around here: 
https://github.com/OPM/opm-simulators/blob/205c4839192d2aa5ced7d6cb88ccede78f246dcd/ebos/eclmpiserializer.hh#L273
Maybe @akva2 knows how to do the nested map serialization.

File: tests/test_GuideRate.cpp
Comment (totto82) [inline tests/test_GuideRate.cpp:523]: This is an example of why I changed the name of the methods and wanted to avoid overloading. "P" is a group but in the original code it got a guiderate target applicable for the wells only. A hard but to notice. I had some similar issues in the wellmodel in opm-simulators.

Comment (joakim-hove) [reply tests/test_GuideRate.cpp:523]: OK - I see.

**Comment by joakim-hove** [comment]
jenkins build this please

File: opm/parser/eclipse/EclipseState/Schedule/Group/GuideRate.hpp
Comment (joakim-hove) [reply opm/parser/eclipse/EclipseState/Schedule/Group/GuideRate.hpp:126]: OK I understand; I don't know how to handle that either.

File: tests/test_GuideRate.cpp
Comment (totto82) [reply tests/test_GuideRate.cpp:523]: I can go back to the "overloading" if you prefer that. Your choice.

**Comment by totto82** [comment]
I notice test failures in cases without injection groups. I will investigate.....

**Comment by totto82** [comment]
jenkins build this opm-simulators=3112 please

**Comment by totto82** [comment]
jenkins build this opm-simulators=3112 please

**Comment by totto82** [comment]
Now only one test case fails and the discrepancy is very small I think a update_data should be fine.

File: tests/test_GuideRate.cpp
Comment (totto82) [reply tests/test_GuideRate.cpp:523]: @joakim-hove Any conclusions on what you prefer here?

Comment (joakim-hove) [reply tests/test_GuideRate.cpp:523]: I prefer overøoad and short names.

Comment (totto82) [reply tests/test_GuideRate.cpp:523]: Ok. Give me some minutes  and I will update accordingly.

**Comment by bska** [comment]
> I think a update_data should be fine.

Please hold off on updating the reference solutions until we've merged OPM/opm-simulators#3039 and OPM/opm-tests#503.

**Comment by totto82** [comment]
jenkins build this opm-simulators=3112 please

**Comment by totto82** [comment]
jenkins build this opm-simulators=3112 update_data please

**Comment by totto82** [comment]
> Please hold off on updating the reference solutions until we've merged OPM/opm-simulators#3039 and OPM/opm-tests#503.

They are merged now. Lets get this in now.

**Comment by totto82** [comment]
jenkins build this opm-simulators=3112 opm-tests=504 please

**Comment by totto82** [comment]
I assume the opm-common part of this is approved. I will merge this and the corresponding part in opm-simulators later today if nobody objects.

**Comment by joakim-hove** [comment]
> I assume the opm-common part of this is approved

yes

**Review by joakim-hove — APPROVED** [review]

**Pull request merged**
